Improve equals check

newly12 opened the pull request:
## What does this PR do?

Improve equals check performance

## Why is it important?

* Improve equals check performance
* improve log output
* fix potentially wrong matches

## Checklist

- [x] My code follows the style guidelines of this project
- [x] ~~I have commented my code, particularly in hard-to-understand areas~~
- [x] ~~I have made corresponding changes to the documentation~~
- [x] ~~I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## How to test this PR locally

Benchmark test

## Logs

Benchmark added test case, before this change
```
pkg: github.com/elastic/beats/v7/libbeat/conditions
BenchmarkEquals_Check
BenchmarkEquals_Check-8   	 1529768	       829 ns/op
PASS
```

After
```
pkg: github.com/elastic/beats/v7/libbeat/conditions
BenchmarkEquals_Check
BenchmarkEquals_Check-8   	 6298933	       188 ns/op
PASS
```

elasticmachine commented:
Pinging @elastic/integrations-services (Team:Services)

urso commented:
Nice!!!

Have you considered to turn `equalsValue` into an interface and do something like this?

```
type equalsValue interface {
  Check(ValuesMap) bool
}

type equalsIntValue int
type equalsStringValue string
...

func (v equalsIntValue) Check(event ValuesMap) bool {
   ...
}
```

alternatively one can try functions as well like:

```
type equalsValue func(ValuesMap) bool

func equalsIntValue(i int) equalsValue {
  return func(event ValuesMap) bool {
    ...
  }
}
```

equalsValue is internal, so we can play with it. Just wondering which implementation would be the fastest here.

All in all I would prefer to move the loop body into a method on `equalsValue`.

newly12 commented:
Thanks for the review. Both look to be more elegant ways to me.

I was trying to implement this and notice that it seems that in both implementation `equalsValue` has to take an `interface{}` as its parameter instead of `ValuesMap`, because the name of the field that is used to extract the value is still saved in `Equals` type. Please correct me if I am wrong, and let me know if I shall move on with the interface parameter.

```
type Equals map[string]equalsValue
```

Another notice is that warning logs will not be logged anymore after this change, since for a field `Equals.check` does not try to extract all supported types from the value but only its wanted type, FYI in case this concerns you.

urso commented:
> Another notice is that warning logs will not be logged anymore after this change, since for a field Equals.check does not try to extract all supported types from the value but only its wanted type, FYI in case this concerns you.

Yes, this is a concern. If checks do not work as expected for all events due to type mismatches, then the logs will point this out.

> I was trying to implement this and notice that it seems that in both implementation equalsValue has to take an interface{} as its parameter instead of ValuesMap, because the name of the field that is used to extract the value is still saved in Equals type.

Ah, you are right. Didn't notice that value is an interface. In that case we would have `type equalsValue interface { Check(value interface{}) (bool, error) }`, right? Then we can log the error if the extraction failed.

newly12 commented:
> Ah, you are right. Didn't notice that value is an interface. In that case we would have type equalsValue interface { Check(value interface{}) (bool, error) }, right? Then we can log the error if the extraction failed.

yeah, I think the difference would be 1. previously warning logs is about all type are mismatched 2. now it warns if the value is not its wanted type. For example, if I have a config `"strField": "foo"`, I got a event with `"strField": 1`, previously there is no warning log, because the value will be extracted as int, though it doesn't equal "foo", now there will be a warning, we expect string but we have a int extracted. 

The result is still the same though.

urso commented:
> now it warns if the value is not its wanted type. For example, if I have a config "strField": "foo", I got a event with "strField": 1, previously there is no warning log, because the value will be extracted as int, though it doesn't equal "foo", now there will be a warning, we expect string but we have a int extracted.

I would say this is an improvement :)

Plus the old code did not take the type in the config in mind when comparing the results. E.g. if `ExtractInt` on the event did succeed, it would have compared the result with the integer value `0`, if the actual setting in the config was a string. This could result in wrong true/false results.

The change will: improve performance, improve log output, fix potentially wrong matches.

newly12 commented:
Hi,

I've added both implementation in the code for now for your reference, the benchmark result is attached below, overall it seems the implementation with new type wrapping int/str/bool has better performance, but I am a bit confused why it shows a lower than others when there are less checks, which I couldn't figure out at this moment..

```
BenchmarkEqualsWith1Conditions
BenchmarkEqualsWith1Conditions-8    	14561558	        77.5 ns/op	       0 B/op	       0 allocs/op
BenchmarkEquals2With1Conditions
BenchmarkEquals2With1Conditions-8   	13009749	        85.4 ns/op	       0 B/op	       0 allocs/op
BenchmarkEquals3With1Conditions
BenchmarkEquals3With1Conditions-8   	14077786	        79.3 ns/op	       0 B/op	       0 allocs/op
BenchmarkEqualsWith3Conditions
BenchmarkEqualsWith3Conditions-8    	 5666799	       205 ns/op	       0 B/op	       0 allocs/op
BenchmarkEquals2With3Conditions
BenchmarkEquals2With3Conditions-8   	 3171140	       384 ns/op	     216 B/op	       4 allocs/op
BenchmarkEquals3With3Conditions
BenchmarkEquals3With3Conditions-8   	 6365893	       195 ns/op	       0 B/op	       0 allocs/op
BenchmarkEqualsWith5Conditions
BenchmarkEqualsWith5Conditions-8    	 2231367	       543 ns/op	     245 B/op	       5 allocs/op
BenchmarkEquals2With5Conditions
BenchmarkEquals2With5Conditions-8   	 3334934	       337 ns/op	     197 B/op	       3 allocs/op
BenchmarkEquals3With5Conditions
BenchmarkEquals3With5Conditions-8   	 3765338	       470 ns/op	     245 B/op	       5 allocs/op
BenchmarkEqualsWith7Conditions
BenchmarkEqualsWith7Conditions-8    	 1584600	       674 ns/op	     315 B/op	       6 allocs/op
BenchmarkEquals2With7Conditions
BenchmarkEquals2With7Conditions-8   	 3286808	       404 ns/op	     221 B/op	       4 allocs/op
BenchmarkEquals3With7Conditions
BenchmarkEquals3With7Conditions-8   	 2253856	       486 ns/op	     245 B/op	       5 allocs/op
```

urso commented:
It looks like the performance depends on the number of allocations. I modified your benchmark function to GC before a run and reset the timer, such that we only test the `Check` itself:

```
func benchmarkEquals(b *testing.B, f factory, fields map[string]interface{}) {
	e, err := f(fields)
	assert.NoError(b, err)

	runtime.GC()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		e.Check(secdTestEvent)
	}
}
```

With this change the numbers for comparisons that did not fail like `BenchmarkEquals.*With1` have been more consistent on my machine, even if I did run the complete suite. for example (without one or the other test did sometimes go up by 20ns on my machine):

```
$ go test -bench 'BenchmarkEquals.*With1' -benchmem
goos: linux
goarch: amd64
pkg: github.com/elastic/beats/v7/libbeat/conditions
BenchmarkEqualsWith1Conditions-32     	17529830	        66.4 ns/op	       0 B/op	       0 allocs/op
BenchmarkEquals2With1Conditions-32    	17563389	        67.6 ns/op	       0 B/op	       0 allocs/op
BenchmarkEquals3With1Conditions-32    	17860872	        65.9 ns/op	       0 B/op	       0 allocs/op
PASS
```

All in all the solution that allocates less will be the fastest: 

```
$ go test -bench 'Benchmark' -benchmem
goos: linux
goarch: amd64
pkg: github.com/elastic/beats/v7/libbeat/conditions
BenchmarkSimpleCondition-32           	59731610	        19.7 ns/op	       0 B/op	       0 allocs/op
BenchmarkCombinedCondition-32         	 9768402	       124 ns/op	       0 B/op	       0 allocs/op
BenchmarkEqualsWith1Conditions-32     	19837894	        58.5 ns/op	       0 B/op	       0 allocs/op
BenchmarkEquals2With1Conditions-32    	18626834	        61.7 ns/op	       0 B/op	       0 allocs/op
BenchmarkEquals3With1Conditions-32    	20692213	        60.5 ns/op	       0 B/op	       0 allocs/op
BenchmarkEqualsWith3Conditions-32     	 6081150	       166 ns/op	       0 B/op	       0 allocs/op
BenchmarkEquals2With3Conditions-32    	 2765434	       570 ns/op	     216 B/op	       4 allocs/op
BenchmarkEquals3With3Conditions-32    	 6134802	       164 ns/op	       0 B/op	       0 allocs/op
BenchmarkEqualsWith5Conditions-32     	 1362459	       885 ns/op	     245 B/op	       5 allocs/op
BenchmarkEquals2With5Conditions-32    	 2274757	       566 ns/op	     221 B/op	       4 allocs/op
BenchmarkEquals3With5Conditions-32    	 1475354	       748 ns/op	     245 B/op	       5 allocs/op
BenchmarkEqualsWith7Conditions-32     	 1000000	      1099 ns/op	     315 B/op	       6 allocs/op
BenchmarkEquals2With7Conditions-32    	 1965396	       656 ns/op	     221 B/op	       4 allocs/op
BenchmarkEquals3With7Conditions-32    	 1529449	       771 ns/op	     245 B/op	       5 allocs/op
BenchmarkNetworkCondition-32          	 3951658	       287 ns/op	      16 B/op	       1 allocs/op
PASS
```

Equals2 was faster when it did allocate once, but for the With3Conditions test it was slower because it has had 4 allocations. Time to profile:

```
$ go test -bench 'BenchmarkEquals2With3Cond' -benchmem -memprofile mem.out -memprofilerate 1 && go tool pprof -http :6060 mem.out
```

This command will open an interactive window in your browser that allows you to inspect the run. In the grapgh click on `equalsBoolValue` to select the node and go to `View -> Source` in the menu. Now we see where the expensive allocation comes from: 

```
    152            .          .           func (e equalsBoolValue) Check(value interface{}) bool { 
    153            .          .           	if bValue, err := ExtractBool(value); err != nil { 
    154            .          .           		return bValue == bool(e) 
    155            .          .           	} 
    156            .    26.75MB           	logp.L().Named(logName).Warnf("expected bool but got type %T in equals condition.", value) 
    157            .          .           	return false 
    158            .          .           } 
```

This is funny. Why does the Equals2 check logs a message and the others not?

The `logp.L().Named(...)` is somewhat expensive. Even more than the Warnf it seems. Idealy we would create the logger in `NewEqualsCondition`, and the store it in the `Equal` type. For Equal2/3 the same. We can either pass the logger as paramater into the check function, or have the check function return an error that gets printed. The `Warnf` statements might still allocated for passing the parameters and for the temporary buffer the log message is written into, before it is written. 

The performance between good path (types match) and bad path (type mismatch) seems to differ quite heavily thanks due to the extra allocations. Looking at the runs without allocations I'd say it doesn't make a big difference which of the three solutions we use, them seem to be similar performance wise:

my run:
```
BenchmarkEqualsWith1Conditions-32     	19837894	        58.5 ns/op
BenchmarkEquals2With1Conditions-32    	18626834	        61.7 ns/op	
BenchmarkEquals3With1Conditions-32    	20692213	        60.5 ns/op
...
BenchmarkEqualsWith3Conditions-32     	 6081150	       166 ns/op
BenchmarkEquals3With3Conditions-32    	 6134802	       164 ns/op
```

a few ns/op difference is negligible and looks like noise.

I'd concentrate on the "good" path only first (just update the benchmark, such that no type errors are printed), select one of these implementations, and finally use the benchmark and memory profile to reduce allocations in the bad path.


Just for fun (equals 3 with allocating the logger upfront per equal function):

```
func equalsIntValue3(i uint64) equalsValueFunc {
	log := logp.L().Named(logName) // <- better to have the logger passed as parameter
	return func(value interface{}) bool {
		if sValue, err := ExtractInt(value); err == nil {
			return sValue == i
		}
		log.Warnf("expected int but got type %T in equals condition.", value)
		return false
	}
}

func equalsStringValue3(s string) equalsValueFunc {
	log := logp.L().Named(logName)
	return func(value interface{}) bool {
		if sValue, err := ExtractString(value); err == nil {
			return sValue == s
		}
		log.Warnf("expected string but got type %T in equals condition.", value)
		return false
	}
}

func equalsBoolValue3(b bool) equalsValueFunc {
	log := logp.L().Named(logName)
	return func(value interface{}) bool {
		if sValue, err := ExtractBool(value); err == nil {
			return sValue == b
		}
		log.Warnf("expected bool but got type %T in equals condition.", value)
		return false
	}
}
```

running the benchmark I get:

```
BenchmarkEqualsWith1Conditions-32     	19428231	        60.5 ns/op	       0 B/op	       0 allocs/op
BenchmarkEquals2With1Conditions-32    	18376104	        63.2 ns/op	       0 B/op	       0 allocs/op
BenchmarkEquals3With1Conditions-32    	17106399	        69.2 ns/op	       0 B/op	       0 allocs/op
BenchmarkEqualsWith3Conditions-32     	 8174004	       150 ns/op	       0 B/op	       0 allocs/op
BenchmarkEquals2With3Conditions-32    	 2119209	       518 ns/op	     216 B/op	       4 allocs/op
BenchmarkEquals3With3Conditions-32    	 7012340	       170 ns/op	       0 B/op	       0 allocs/op
BenchmarkEqualsWith5Conditions-32     	 2335826	       873 ns/op	     245 B/op	       5 allocs/op
BenchmarkEquals2With5Conditions-32    	 3561202	       573 ns/op	     221 B/op	       4 allocs/op
BenchmarkEquals3With5Conditions-32    	 3564328	       443 ns/op	      56 B/op	       1 allocs/op
BenchmarkEqualsWith7Conditions-32     	 1000000	      1087 ns/op	     315 B/op	       6 allocs/op
BenchmarkEquals2With7Conditions-32    	 2052055	       541 ns/op	     186 B/op	       3 allocs/op
BenchmarkEquals3With7Conditions-32    	 2262382	       526 ns/op	      56 B/op	       1 allocs/op
```

The single allocation is due the logger doing  `fmt.Sprintf` and can't be removed.

newly12 commented:
oops this is my fault.. I debugged equals2 finally got why it hits the logger statement in 3Condition test of equals2, it should be `err == nil` here. 🤦‍♂️

```
	if bValue, err := ExtractBool(value); err != nil {
```

and after that I am getting expected output

```
BenchmarkEqualsWith1Conditions
BenchmarkEqualsWith1Conditions-8    	14012263	        75.6 ns/op	       0 B/op	       0 allocs/op
BenchmarkEquals2With1Conditions
BenchmarkEquals2With1Conditions-8   	15052548	        79.3 ns/op	       0 B/op	       0 allocs/op
BenchmarkEquals3With1Conditions
BenchmarkEquals3With1Conditions-8   	15445071	        77.9 ns/op	       0 B/op	       0 allocs/op
BenchmarkEqualsWith3Conditions
BenchmarkEqualsWith3Conditions-8    	 5918541	       190 ns/op	       0 B/op	       0 allocs/op
BenchmarkEquals2With3Conditions
BenchmarkEquals2With3Conditions-8   	 6046137	       202 ns/op	       0 B/op	       0 allocs/op
BenchmarkEquals3With3Conditions
BenchmarkEquals3With3Conditions-8   	 6471944	       188 ns/op	       0 B/op	       0 allocs/op
BenchmarkEqualsWith5Conditions
BenchmarkEqualsWith5Conditions-8    	 2222293	       553 ns/op	     245 B/op	       5 allocs/op
BenchmarkEquals2With5Conditions
BenchmarkEquals2With5Conditions-8   	 3954058	       460 ns/op	     245 B/op	       5 allocs/op
BenchmarkEquals3With5Conditions
BenchmarkEquals3With5Conditions-8   	 2304970	       462 ns/op	     245 B/op	       5 allocs/op
BenchmarkEqualsWith7Conditions
BenchmarkEqualsWith7Conditions-8    	 1786755	       683 ns/op	     315 B/op	       6 allocs/op
BenchmarkEquals2With7Conditions
BenchmarkEquals2With7Conditions-8   	 2730859	       500 ns/op	     245 B/op	       5 allocs/op
BenchmarkEquals3With7Conditions
BenchmarkEquals3With7Conditions-8   	 2750624	       468 ns/op	     245 B/op	       5 allocs/op
```

Now seems equals3 is the choice, and as you have mentioned, `logp.L().Named(...)` is expensive, the way to avoid keeping getting the logger would be get it once at the time getting the "equals" instance. 

I have updated the code to adopt equals3 and cleaned other code. Please have a review. 

It's been very helpful, I really appreciate it!

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > The PR is not allowed to run in the CI yet
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-22778/detail/PR-22778/2//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-22778/detail/PR-22778/2//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-22778/detail/PR-22778/2//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-22778/detail/PR-22778/2//artifacts) 

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `Pull request #22778 updated`

* Reason: `The PR is not allowed to run in the CI yet`

* Start Time: 2020-12-05T05:03:11.485+0000

* Duration: 3 min 25 sec


* Commit: c56e29ad6f55b2a6f4a1e50e1e565fe2105b3b50




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->





### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-22778/detail/PR-22778/2//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>

  
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-22778/runs/2/steps/83/log/?start=0">here</a></li>
<li>Description: <code>githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permission</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-12-05T05:04:52.242Z]  > git clean -fdx # timeout=10
[2020-12-05T05:04:52.254Z] Fetching upstream changes from git@github.com:elastic/beats.git
[2020-12-05T05:04:52.254Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-12-05T05:04:52.257Z]  > git fetch --no-tags --progress --prune -- git@github.com:elastic/beats.git +refs/pull/22778/head:refs/remotes/origin/PR-22778 +refs/heads/master:refs/remotes/origin/master # timeout=15
[2020-12-05T05:04:53.303Z]  > git config core.sparsecheckout # timeout=10
[2020-12-05T05:04:53.307Z]  > git checkout -f c56e29ad6f55b2a6f4a1e50e1e565fe2105b3b50 # timeout=15
[2020-12-05T05:04:54.757Z] Merge succeeded, producing c56e29ad6f55b2a6f4a1e50e1e565fe2105b3b50
[2020-12-05T05:04:54.757Z] Checking out Revision c56e29ad6f55b2a6f4a1e50e1e565fe2105b3b50 (PR-22778)
[2020-12-05T05:04:54.731Z]  > git remote # timeout=10
[2020-12-05T05:04:54.735Z]  > git config --get remote.origin.url # timeout=10
[2020-12-05T05:04:54.738Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-12-05T05:04:54.742Z]  > git merge 02cf5643f3c9d82088f6ef8d32a1f46e656ba00c # timeout=10
[2020-12-05T05:04:54.752Z]  > git rev-parse HEAD^{commit} # timeout=10
[2020-12-05T05:04:54.758Z]  > git config core.sparsecheckout # timeout=10
[2020-12-05T05:04:54.762Z]  > git checkout -f c56e29ad6f55b2a6f4a1e50e1e565fe2105b3b50 # timeout=15
[2020-12-05T05:04:59.343Z] Commit message: "update CHANGELOG.next.asciidoc"
[2020-12-05T05:04:59.343Z] First time build. Skipping changelog.
[2020-12-05T05:04:59.343Z] Cleaning workspace
[2020-12-05T05:04:59.344Z]  > git rev-parse --verify HEAD # timeout=10
[2020-12-05T05:04:59.349Z] Resetting working tree
[2020-12-05T05:04:59.349Z]  > git reset --hard # timeout=10
[2020-12-05T05:04:59.429Z]  > git clean -fdx # timeout=10
[2020-12-05T05:05:00.403Z] Timeout set to expire in 3 hr 0 min
[2020-12-05T05:05:00.427Z] The timestamps step is unnecessary when timestamps are enabled for all Pipeline builds.
[2020-12-05T05:05:00.787Z] [INFO] Number of builds to be searched 10
[2020-12-05T05:05:01.368Z] [INFO] 'shallow' is forced to be disabled when running on PullRequests
[2020-12-05T05:05:01.395Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-22778/src/github.com/elastic/beats
[2020-12-05T05:05:01.428Z] [INFO] gitCheckout: Checkout SCM PR-22778 with default customisation from the Item.
[2020-12-05T05:05:01.456Z] [INFO] Override default checkout
[2020-12-05T05:05:01.515Z] Sleeping for 10 sec
[2020-12-05T05:05:11.590Z] using credential f6c7695a-671e-4f4f-a331-acdce44ff9ba
[2020-12-05T05:05:11.641Z] Wiping out workspace first.
[2020-12-05T05:05:11.649Z] Cloning the remote Git repository
[2020-12-05T05:05:11.649Z] Using shallow clone with depth 10
[2020-12-05T05:05:11.649Z] Avoid fetching tags
[2020-12-05T05:05:11.660Z] Cloning repository git@github.com:elastic/beats.git
[2020-12-05T05:05:11.680Z]  > git init /var/lib/jenkins/workspace/Beats_beats_PR-22778/src/github.com/elastic/beats # timeout=10
[2020-12-05T05:05:11.685Z] Fetching upstream changes from git@github.com:elastic/beats.git
[2020-12-05T05:05:11.685Z]  > git --version # timeout=10
[2020-12-05T05:05:11.688Z]  > git --version # 'git version 2.17.1'
[2020-12-05T05:05:11.689Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-12-05T05:05:11.692Z]  > git fetch --no-tags --progress -- git@github.com:elastic/beats.git +refs/heads/*:refs/remotes/origin/* # timeout=15
[2020-12-05T05:05:29.024Z] Cleaning workspace
[2020-12-05T05:05:29.035Z] Using shallow fetch with depth 10
[2020-12-05T05:05:29.035Z] Pruning obsolete local branches
[2020-12-05T05:05:29.010Z]  > git config remote.origin.url git@github.com:elastic/beats.git # timeout=10
[2020-12-05T05:05:29.013Z]  > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
[2020-12-05T05:05:29.020Z]  > git config remote.origin.url git@github.com:elastic/beats.git # timeout=10
[2020-12-05T05:05:29.026Z]  > git rev-parse --verify HEAD # timeout=10
[2020-12-05T05:05:29.028Z] No valid HEAD. Skipping the resetting
[2020-12-05T05:05:29.028Z]  > git clean -fdx # timeout=10
[2020-12-05T05:05:29.037Z] Fetching upstream changes from git@github.com:elastic/beats.git
[2020-12-05T05:05:29.037Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-12-05T05:05:29.040Z]  > git fetch --no-tags --progress --prune -- git@github.com:elastic/beats.git +refs/pull/22778/head:refs/remotes/origin/PR-22778 +refs/heads/master:refs/remotes/origin/master # timeout=15
[2020-12-05T05:05:30.065Z] Merging remotes/origin/master commit 02cf5643f3c9d82088f6ef8d32a1f46e656ba00c into PR head commit c56e29ad6f55b2a6f4a1e50e1e565fe2105b3b50
[2020-12-05T05:05:30.067Z]  > git config core.sparsecheckout # timeout=10
[2020-12-05T05:05:30.070Z]  > git checkout -f c56e29ad6f55b2a6f4a1e50e1e565fe2105b3b50 # timeout=15
[2020-12-05T05:05:31.450Z] Merge succeeded, producing c56e29ad6f55b2a6f4a1e50e1e565fe2105b3b50
[2020-12-05T05:05:31.450Z] Checking out Revision c56e29ad6f55b2a6f4a1e50e1e565fe2105b3b50 (PR-22778)
[2020-12-05T05:05:32.183Z] Commit message: "update CHANGELOG.next.asciidoc"
[2020-12-05T05:05:32.183Z] Cleaning workspace
[2020-12-05T05:05:31.429Z]  > git remote # timeout=10
[2020-12-05T05:05:31.432Z]  > git config --get remote.origin.url # timeout=10
[2020-12-05T05:05:31.435Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-12-05T05:05:31.438Z]  > git merge 02cf5643f3c9d82088f6ef8d32a1f46e656ba00c # timeout=10
[2020-12-05T05:05:31.446Z]  > git rev-parse HEAD^{commit} # timeout=10
[2020-12-05T05:05:31.452Z]  > git config core.sparsecheckout # timeout=10
[2020-12-05T05:05:31.457Z]  > git checkout -f c56e29ad6f55b2a6f4a1e50e1e565fe2105b3b50 # timeout=15
[2020-12-05T05:05:32.185Z]  > git rev-parse --verify HEAD # timeout=10
[2020-12-05T05:05:32.188Z] Resetting working tree
[2020-12-05T05:05:32.188Z]  > git reset --hard # timeout=10
[2020-12-05T05:05:32.258Z]  > git clean -fdx # timeout=10
[2020-12-05T05:05:33.326Z] Masking supported pattern matches of $GIT_USERNAME or $GIT_PASSWORD
[2020-12-05T05:05:33.915Z] + git fetch https://****:****@github.com/elastic/beats.git +refs/pull/*/head:refs/remotes/origin/pr/*
[2020-12-05T05:06:30.268Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-22778/src/github.com/elastic/beats/.git
[2020-12-05T05:06:30.305Z] Archiving artifacts
[2020-12-05T05:06:31.014Z] + git rev-parse HEAD
[2020-12-05T05:06:31.368Z] + git rev-parse HEAD
[2020-12-05T05:06:31.672Z] + git rev-parse origin/pr/22778
[2020-12-05T05:06:31.719Z] [INFO] githubEnv: Found Git Build Cause: pr
[2020-12-05T05:06:32.083Z] Masking supported pattern matches of $GITHUB_TOKEN
[2020-12-05T05:06:33.088Z] [WARN] githubApiCall: The REST API call https://api.github.com/repos/elastic/beats/pulls/22778/reviews return 0 elements
[2020-12-05T05:06:33.144Z] [INFO] githubPrCheckApproved: Title: Improve equals check performance - User: newly12 - Author Association: CONTRIBUTOR
[2020-12-05T05:06:33.568Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet
[2020-12-05T05:06:33.568Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permissions can do so.)
[2020-12-05T05:06:33.627Z] [INFO] Let's stop build #2. The PR is not allowed to run in the CI yet
[2020-12-05T05:06:33.649Z] Sleeping for 5 sec
[2020-12-05T05:06:35.244Z] Stage "Lint" skipped due to earlier failure(s)
[2020-12-05T05:06:35.319Z] Stage "Build&Test" skipped due to earlier failure(s)
[2020-12-05T05:06:35.393Z] Stage "Packaging" skipped due to earlier failure(s)
[2020-12-05T05:06:35.493Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-22778/src/github.com/elastic/beats
[2020-12-05T05:06:36.013Z] Running on Jenkins in /var/lib/jenkins/workspace/Beats_beats_PR-22778
[2020-12-05T05:06:36.290Z] [INFO] getVaultSecret: Getting secrets
[2020-12-05T05:06:36.374Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-12-05T05:06:37.200Z] + chmod 755 generate-build-data.sh
[2020-12-05T05:06:37.200Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-22778/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-22778/runs/2 ABORTED 205445
[2020-12-05T05:06:37.200Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-22778/runs/2/steps/?limit=10000 -o steps-info.json
[2020-12-05T05:06:37.751Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-22778/runs/2/tests/?status=FAILED -o tests-errors.json
[2020-12-05T05:06:38.001Z] Retry 1/3 exited 22, retrying in 1 seconds...
[2020-12-05T05:06:38.912Z] Retry 2/3 exited 22, retrying in 2 seconds...

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

urso commented:
jenkins run the tests please

urso commented:
The change looks good to me.

Thank you, this is a really great improvement in the hot code paths!

Do you by any chance have a benchmark run (with the new benchmarks) for the old code? I'd love to see how much performance has improved. You might want to check out benchcmp (e.g. see [this blog post](https://pliutau.com/measure-performance-changes/), not by me, just a random google search)

I started CI and will merge and port the change to the release branches if CI goes green.

urso commented:
Merged.

Thanks for the contribution! Performance improvements like this are really exciting!

newly12 commented:
Thanks @urso . I had compared twice since the result from some other benchmark seems confusing or misleading, the result for equals check is consistent.  

```
$ benchcmp old.txt new.txt   
benchcmp is deprecated in favor of benchstat: https://pkg.go.dev/golang.org/x/perf/cmd/benchstat
benchmark                          old ns/op     new ns/op     delta
BenchmarkSimpleCondition-8         47.6          24.7          -48.11%
BenchmarkCombinedCondition-8       134           221           +64.93%
BenchmarkEquals/1_condition-8      336           79.3          -76.40%
BenchmarkEquals/3_conditions-8     824           203           -75.36%
BenchmarkEquals/5_conditions-8     1285          315           -75.49%
BenchmarkEquals/7_conditions-8     917           354           -61.40%
BenchmarkNetworkCondition-8        173           171           -1.16%

$ benchcmp old2.txt new2.txt 
benchcmp is deprecated in favor of benchstat: https://pkg.go.dev/golang.org/x/perf/cmd/benchstat
benchmark                          old ns/op     new ns/op     delta
BenchmarkSimpleCondition-8         22.5          35.6          +58.22%
BenchmarkCombinedCondition-8       136           141           +3.68%
BenchmarkEquals/1_condition-8      432           73.6          -82.96%
BenchmarkEquals/3_conditions-8     786           194           -75.32%
BenchmarkEquals/5_conditions-8     1224          315           -74.26%
BenchmarkEquals/7_conditions-8     842           278           -66.98%
BenchmarkNetworkCondition-8        165           180           +9.09%
```
